### ISSUE TITLE
Road to 3.0

### ISSUE THREAD

**Issue by pierrejoye** [libgd/libgd#705]
Hi Team,

I started a little wiki page for the roadmap to 3.0: [https://github.com/libgd/libgd/wiki/3.0-Roadmap](url). 

There is more to add I am sure (supported builds script f.e.) but these are the ones I have been struggling to finish since too long :)

**Comment by cmb69** [comment]
I'd like to drop support for explicit palette images. It's painful to maintain (we had some sec issues due to confusing `pixels` and `tpixels`, and `transparent` complicates things). And it's hard for users to get this right, if antialising is used, due to exhaustion of the palette entries. That would be a pretty big BC break, though.

Regarding the build scripts: I think we should drop everything besides autotools and CMake.

**Comment by pierrejoye** [comment]
I tend to agree but still useful for some users with a load of users
(embedded devices with lcd display).

However we could just freeze it and leave it as is.

As of the new images types, I would like a new struct, cleaner, continuous
buffer, a stride and a format along with minimal meta if needed.

Anything else can be their own struct/class.

**Comment by pierrejoye** [comment]
![out-gray](https://user-images.githubusercontent.com/282408/124374007-b9dd0280-dcc1-11eb-8a82-59ad111eafa3.png)
Example of Freetype Raster usage to a GD TR image. It is using this shape, [https://www.freetype.org/freetype2/docs/tutorial/example3.cpp](url) but on a GD buffer. Also quality will be affected using alpha channel as AA given we only have 127 levels (ARGB32 solves this).

**Comment by pierrejoye** [comment]
Updated a bit adding proposals for APIs and more details. Please feel free to add your thoughts and concepts as well :)

**Comment by pierrejoye** [comment]
Pushed beginning of Surface support in features/surface. Feedback welcome. See the Wiki link for what is done or not (will update accordingly).

**Comment by pierrejoye** [comment]
LineTo MoveTo LineWidth done

Getting there :)

I will commit over the weekend, need some cleanup

![filled](https://user-images.githubusercontent.com/282408/125037477-6520f980-e0be-11eb-94e0-02bfe9abb2a3.png)
![stroke - Copy](https://user-images.githubusercontent.com/282408/125037509-70742500-e0be-11eb-98d3-568197848e0f.png)
![stroke](https://user-images.githubusercontent.com/282408/125037533-766a0600-e0be-11eb-9e0e-54d67d92008c.png)

**Comment by pierrejoye** [comment]
And simple blending works (respecting alpha):

![image](https://user-images.githubusercontent.com/282408/125068710-fd7ba600-e0df-11eb-977b-91fc21b3155d.png)

**Comment by pierrejoye** [comment]
Ok, almost there. Using the famous tiger:

![image](https://user-images.githubusercontent.com/282408/125162530-8ebb4d00-e1b2-11eb-8cc3-97dd0285abfd.png)

Clipping and then 0.1 is ready.

**Comment by pierrejoye** [comment]
OK, done, took a bit more hours this weekend, but almost everything in place. Cleanup and commit today.

![tigerclipped](https://user-images.githubusercontent.com/282408/125259041-2642ac00-e329-11eb-9917-b16fc47b7878.png)
![pathclip2](https://user-images.githubusercontent.com/282408/125259059-2b9ff680-e329-11eb-8bab-f53eea3d92c3.png)

**Comment by pierrejoye** [comment]
![curvedrectangle](https://user-images.githubusercontent.com/282408/125259148-407c8a00-e329-11eb-96ba-cba8468df440.png)

**Comment by pierrejoye** [comment]
Code has been pushed.

I only update CMake, please use it to build or test. I apologize for the WS changes, I realize it too late (messed a few headers with WS changes :().

https://github.com/libgd/libgd/tree/features/surface

**Comment by cmb69** [comment]
> Code has been pushed.

Very interesting, but I won't have much time time to take a closer look, unfortunately.

Do I understand correctly that `gdSurface` would be an alternative to `gdImage` without any interopability? Is support for other image formats than PNG missing yet, or is it not planned at all? Is it supposed to obsolete the existing drawing primitives in the long run? What about the performance?

**Comment by pierrejoye** [comment]
hello,

gdSurface will be what gdImage should but can't without breaking BC very badly.

However I will add then all functions to support the new formats via new APIs. Doing so binding can simply keep BC by switching the calls. All image formats will be supported as long as it is possible.

Performance wise, true 32bits pixels stored in a continuous buffer are way better to handle than what we have now (31bits with some non standard way to handle alpha or antialiasing (-color). Things like SSE2-4 are possible (blending using 64/128 or more bits in one cpu op f.e.).

Another good side is that these formats work 100% fine with cairo, gtk, windows, and almost any other libs out there. So interop will be very easy.

For now I focus on GD 2D functions with true vector based rendering (like cairo or agg). This is missing so badly in libgd as the current functions are really too 90s 2000s. Once it is, API wise, done, I will attack the encoder/decoder to add support for all file formats, then comes the other functions like filters and co.

I am not sure about imagecircle etc as it will break BC from a rendering point of view, mostly for users having created pixel exact charts, that will break it up (gdDraw is subpixels but exact).

At some point I will add some support for another 2D library with JIT renderer, sadly only x86, but it is the fastest software only rasterer I know. This is where libgd is used, web, embedded with low or not GPU support.

**Comment by cmb69** [comment]
Thanks for the clarifications! This sounds very interesting, but is certainly a lot of work. I hope we can mange that. :)

I'm not that much concerned regarding some minor rendering differences, if we do that in a major release, and support 2.x for some time after that release.

**Comment by pierrejoye** [comment]
You are welcome :)

And yes it is. The new features are the _easy_ ;-) parts. Making the other compatible is easy but takes time. Especially the old drawing functions. Filters and co are relatively easy as I have them all for my old never released libpimp (which has 32bit format already ;).

**Comment by pierrejoye** [comment]
@cmb69 something like this would be great btw, or maybe rely on it optionally https://github.com/ermig1979/Simd 
Works for intel, amd, powerpc and ARM/Neon

**Comment by cmb69** [comment]
Interesting. I'm still a bit concerned, though, that basically rewriting libgd might take ages given our current bandwidth.

**Comment by pierrejoye** [comment]
On Thu, Jul 15, 2021 at 11:58 PM Christoph M. Becker <
***@***.***> wrote:

> Interesting. I'm still a bit concerned, though, that basically rewriting
> libgd might take ages given our current bandwidth.
>
Also there is no choice to move forward. The current internal formats are
useless in real world and as you know, create more issues (transparency,
weird rules, etc). It basically prevents GD to move forward or to be used
with modern 3rd parties libraries or applications.

The surface part should be in the couple of fix. Really not difficult to
add the various image formats coder/decoders. The filters and
transformation functions take some time but are straightforward. My only
worries are really about the old drawing functions, what to do with them.

What do you think?


Best,
-- 
Pierre

@pierrejoye | http://www.libgd.org

**Comment by pierrejoye** [comment]
Last but not least, we are in no hurry :)

**Comment by pierrejoye** [comment]
You can follow the progresses of gdDraw here https://github.com/libgd/libgd/wiki/GD-Draw,-vector-based-drawing-APIs

**Comment by ChrisBarker-NOAA** [comment]
Not sure if I should create a new issue for this, but for now, a couple notes:

Back in the day, I looked into Cairo for a bit -- this may be outdated, but I found two issues that I couldn't figure out a good way to work around:

1) No ellipse support -- the way to draw an ellipse was to draw a circle and transform it. That would be perfectly fine except for the stroke: the transform would transform the stroke width as well, so that there was literally no way to draw an ellipse with a consistent stroke width.

2) stroke and fill independent: for much rendering, that works fine, you fill the path, and then draw the stroke on top of it -- looks great. However, if you want a semi-transparent object, then it gets odd, because the semi-transparent stroke shows the underlying fill. I think the only solution to this was to create another surface, render the object opaque, then blit the  temporary surface onto the main surface applying the transparency. It. may be that that's what has to happen under the hood, but it would be really nice for the API to simply support drawing an object with a fill and stroke that "act" as one thing.

**Comment by pierrejoye** [comment]
do you mean what is described here?

https://www.cairographics.org/cookbook/ellipses/

**Comment by ChrisBarker-NOAA** [comment]
@pierrejoye: yes, that's it - -and a solution too! I'm pretty sure that solution was not available ten years ago :-)
